Integrate creation of chocolatey packages

CharliePoole opened the pull request:
Fixes #8 and will be a model for the other extensions as well.

I've added comments inline to explain some choices I made. @nunit/core-team Please take a look as this will be the basis for the other extensions as well as the runners.

Also fixes #2

CharliePoole committed:
First shot at integrating chocolatey package creation

CharliePoole committed:
Update package references and remove output files from repository

CharliePoole commented on build.cake:
```diff
@@ -13,6 +12,9 @@ var UNIT_TEST_ASSEMBLY = "nunit-project-loader.tests.dll";
 var configuration = Argument("configuration", "Debug");
+var nugetVersion = Argument("nugetVersion", (string)null);
+var chocoVersion = Argument("chocoVersion", (string)null);
+var binaries = Argument("binaries", (string)null);
```
Optional args used by the Package target. The two version args override the package version that we otherwise calculate. The `binaries` arg causes the packaging to be made from a directory other than BIN_DIR. This would have to have the desired binaries in it, from an external source. I've tested the args and they work, but I'm not yet seeing how we would get the binaries for a repackaging situation. Ideas?

ChrisMaddock replied:
Download and unpack a NuGet package or whatever? Eventually, we could release from an Appveyor build or something. I think repackaging is rare enough a situation that it can require a little manual intervention.

CharliePoole replied:
That was how I did the first packages (in a separate repo) but I think we may want to be able to repackage for nuget as well as chocolatey. Also, one reason you repackage is because you forgot a binary in the original package. I guess it may have to be manual.

CharliePoole commented on build.cake:
```diff
@@ -116,11 +122,24 @@ Task("Build")
+				.WithProperty("Configuration", configuration)
+				.SetVerbosity(Verbosity.Minimal)
+			);
+		}
```
I get constant errors with DotNetBuild, so I went to the approach we use for the framework.

CharliePoole commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
 	.Does(() =>
 	{
-		NUnit3(UNIT_TEST_PATH);
+		NUnit3(BIN_DIR + UNIT_TEST_ASSEMBLY);
```
Eliminating a one-time-use constant, here and elsewhere. These are actually legacy from using `nant`.

CharliePoole commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
+				Files = new [] {
+					new ChocolateyNuSpecContent { Source = "../LICENSE.txt", Target = "tools" },
+					new ChocolateyNuSpecContent { Source = "VERIFICATION.txt", Target = "tools" },
+					new ChocolateyNuSpecContent { Source = binaries + "nunit-project-loader.dll", Target = "tools" }
```
Note: everything should go into the tools directory for chocolatey, because they don't copy files in the root when install the package.

CharliePoole commented on nuget/nunit-project-loader.nuspec:
```diff
@@ -17,8 +17,4 @@
-  <files>
-    <file src="../../LICENSE.txt" />
-    <file src="nunit-project-loader.dll" target="tools"/>
-  </files>
```
Removing files and leaving only metadata in the nuspec puts both chocolatey and nuget on the same footing. Files to be included are supplied by the script.

ChrisMaddock commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
+				Version = nugetVersion ?? packageVersion,
+				OutputDirectory = OUTPUT_DIR,
+				Files = new [] {
+					new NuSpecContent { Source = "../LICENSE.txt" },
```
Is this the right path for Licence.txt? (https://github.com/nunit/nunit-project-loader/issues/7) Surprised Cake doesn't flag the missing file. 😞 

It would be good to have this as a rooted path for the avoidance of doubt, e.g. `Path.Combine(PROJECT_DIR, LICENSE.txt)`

CharliePoole replied:
Yes, files specified in this way are relative to the location of the nuspec file. Originally, all paths were absolute with many more constants. I wanted to make it lighter weight and simpler, but it sounds as if it's a little confusing. OTOH, I really like using just the filename for those things that are in the same directory as the choco file, as it makes it clear (at least to me) that that's where they are. 

Let's see what others think since it's purely a matter of style.

ChrisMaddock commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
+				Version = chocoVersion ?? packageVersion,
+				OutputDirectory = OUTPUT_DIR,
+				Files = new [] {
+					new ChocolateyNuSpecContent { Source = "../LICENSE.txt", Target = "tools" },
```
As above on the licence file path

ChrisMaddock reviewed (approved):
Looks good - just the file path changes below.

One more I can't comment on directly - you seem to have accidentally added a couple of binary files, `choco/nunit-extension-nunit-project-loader.3.6.0.nupkg` and `nuget/NUnit.Extension.NUnitProjectLoader.3.6.0.nupkg`.

ChrisMaddock commented:
I added 'fixes #2' to your opening post, and hopefully this will also fix #7 once the paths are corrected. 😄

ChrisMaddock commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
+	.IsDependentOn("Build")
+	.Does(() =>
+	{
+		ChocolateyPack("choco/nunit-project-loader.choco.nuspec",
```
Have you noticed the nupkg isn't taking the `.choco.` suffix? Looks like it's uses the package id + version instead?

Maybe best to file rename after packing to get this, if you wanted it? @ferventcoder suggested that wasn't good for choco on https://github.com/nunit/nunit-console/issues/253#issuecomment-315315820 however, maybe best to see how that gets resolved first.

If the solution is not to do it, it would be good to remove the `.choco` from the nuspec filename.

CharliePoole replied:
The name of the output file from NuGet Pack is independent of the name of the input file! They use the ID that is given in the file. So any renaming has to be separate and is not yet implemented while we discuss the pros and cons.

ChrisMaddock commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
 
-Task("Package")
+Task("PackageNuGet")
 	.IsDependentOn("Build")
```
This will always run build before packaging. I presume you meant to remove this?

CharliePoole replied:
No... or at least not yet... This gets to the question of how we access pure repackaging. Do we want a Repackage target that doesn't depend on build? Should Package depend on Build and Repackage? etc. I elected not to pick one but to see what folks thought first.

Right now, if you want to repackage the chocolatey package, you have to enter a pretty obscure command...

```
./build.ps1 -t Package -ScriptArgs "--chocoVersion=3.5.1 --binaries=Path/to/binaries"
```

You will trigger a build but the output will not be used for chocolatey. Instead, the path you specify will provide the binaries. 

Even if we don't do it often, I think it would be nice to provide a friendlier front end for this.

ChrisMaddock commented on build.cake:
```diff
@@ -131,25 +150,44 @@ Task("Test")
+	});
+
+Task("PackageChocolatey")
+	.IsDependentOn("Build")
```
As above.

CharliePoole commented:
Apparently our standard .gitignore file doesn't have .nupkg in it.

CharliePoole commented:
@ChrisMaddock Here's an idea...

* Package depends on Build, PackageNuGet, PackageChocolatey
* PackageNuGet and PackageChoco no longer depend on build
* Normal builds use Package as the target
* If we are repackaging, we probably only want to do it for one of the two packages, so we set up the directory containg binaries and do either PackageNuGet or PackageChocolatey.
* We could also add a dependent task to the PackageTarget that checks the arguments. The `binaries` argument should be null if we are running Package.

CharliePoole committed:
Remove nupkg files accidentally commitede

ChrisMaddock commented:
That all seems sensible.

To be honest, I'm not too worried here. I feel like repackaging is an unusual enough task, it's ok for it to include some manual steps. I just thought it was your intention, to have the two stages separate again, like we used to.

CharliePoole commented:
@ChrisMaddock I would prefer to have them separate. Others in the @nunit/core-team have asked for the dependency to be retained - at least as I understood the comments.

With the dependency, whoever releases just runs the Package target. The extension is built and the result is packaged. Repackaging still works, by using PackageNuGet or PackageChocolatey

Without the dependency, they have to
1. Run the build or test target or otherwise verify that the binaries are up to date
2. Run the package command
Repackage works the same as without the dependency.

@nunit/core-team What is your pleasure? I think it will be best if we eventually make all our build scripts work in whichever way we decide.

CharliePoole committed:
Do the entire packaging in the script, without a nuspec file

CharliePoole commented:
New commit - @nunit/core-team I'd appreciate a re-review.

* No more nuspecs - it's all done in build.cake
* Changed dependencies and added a RePackage target
* Error if you use Build or Package with the --binaries option

Further suggestions?

CharliePoole commented:
Looks like I lost my CreateDirectory command.

CharliePoole committed:
Create directory if not present

CharliePoole committed:
Restructure for ease of replication and upgrade nunit console runner

CharliePoole commented:
I think this should now be ready to merge. Most recently, I've been working offline at creating the same script for the vs-project-loader with the result that I reorganized so the items I needed to change are in one section at the top of the script. 

Also updated to run tests with latest console runner.

CharliePoole committed:
Refactoring
